src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,227 @@
+/**
+ * @file cne_impl.hpp
+ * @author Marcus Edel
+ * @author Kartik Nighania
+ *
+ * Conventional Neural Evolution
+ * An optimizer that works like biological evolution which selects best
+ * candidates based on their fitness scores and creates new generation by
+ * mutation and crossover of population.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+
+#include "cne.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+CNE::CNE(const size_t populationSize,
+         const size_t maxGenerations,
+         const double mutationProb,
+         const double mutationSize,
+         const double selectPercent,
+         const double tolerance,
+         const double objectiveChange) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    mutationProb(mutationProb),
+    mutationSize(mutationSize),
+    selectPercent(selectPercent),
+    tolerance(tolerance),
+    objectiveChange(objectiveChange),
+    numElite(0),
+    elements(0)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename DecomposableFunctionType>
+double CNE::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4)
+    throw std::logic_error("Population size should be atleast 4!");
+
+  // Find the number of elite canditates from population.
+  numElite = floor(selectPercent * populationSize);
+
+  // Making sure we have even number of candidates to remove and create.
+  if ((populationSize - numElite) % 2 != 0) numElite--;
+
+  // Terminate if two parents can not be created.
+  if (numElite < 2)
+    throw std::logic_error("Increase selection percentage.");
```

Comment:
I think this could be a better error message---the comment above might help it turn into something more useful.  With this error the user just gets `Increase selection percentage.` and they aren't sure why.  I'd suggest something more like:

```
throw std::logic_error("Selection percentage is too low---cannot select two parents.  Increase selection percentage.");
```

This idea could probably be applied to a few more error messages here.